File: zlcore/src/main/java/com/zaitunlabs/zlcore/utils/FormValidationUtils.java
package com.zaitunlabs.zlcore.utils;

import android.app.Activity;
import android.content.Context;
import android.text.TextUtils;
import android.util.Patterns;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by ahsai on 11/21/2017.
 */

public class FormValidationUtils {
    public static final int DefaultType = 1;
    public static final int IdleType = 2;
    public static final int UnfocusType = 3;
    private Context context;
    private List<Validator> mValidatorList;

    public FormValidationUtils(Context context){
        this.context = context.getApplicationContext();
        mValidatorList = new ArrayList<>();
    }
    public void addValidator(Validator validator){
        addValidator(validator,DefaultType);
    }

    public void addValidator(Validator validator, int validationType){
        mValidatorList.add(validator);
        if(validationType == DefaultType) {
        } else if(validationType == IdleType){
            enableTypingValidation(validator);
        } else if(validationType == UnfocusType){
            enableUnfocusValidation(validator);
        }
    }

    private void enableUnfocusValidation(final Validator validator) {
        CommonUtils.performTaskWhenUnFocus(validator.mView, new Runnable() {
            @Override
            public void run() {
                validator.validate();
            }
        });
    }

    private void enableTypingValidation(final Validator validator){
        if(validator.mView instanceof EditText) {
            CommonUtils.performTaskWhenTypeIdle((EditText) validator.mView, new Runnable() {
                @Override
                public void run() {
                    validator.validate();
                }
            });
        }
    }

    public Validator getValidator(int position) {
        if(position < mValidatorList.size()) {
            return mValidatorList.get(position);
        }
        return null;
    }

    public int getValidatorCount(){
        return mValidatorList.size();
    }

    public void removeValidator(Validator validator){
        mValidatorList.remove(validator);
    }

    public boolean validate() throws ValidatorException {
        boolean isFirsErrorFocused = false;
        for (int x=0; x<mValidatorList.size(); x++){
            Validator validator = mValidatorList.get(x);
            for (int y=0; y<validator.mRuleList.size(); y++){
                AbstractValidatorRule rule = validator.mRuleList.get(y);
                if(validator.mView instanceof TextView) {
                    TextView mTextView = (TextView) validator.mView;
                    boolean isValid = rule.isValid(validator.mView, mTextView.getText().toString(), validator.packet);
                    if (!isValid) {
                        mTextView.setError(rule.getMessage());
                        if (!isFirsErrorFocused) {
                            mTextView.requestFocus();
                            isFirsErrorFocused = true;
                        }
                        break;
                    } else {
                        mTextView.setError(null);
                    }
                }
            }
        }

        return !isFirsErrorFocused;
    }

    public boolean validate(OnValidationCallback onValidationCallback) throws ValidatorException {
        int totalRule = 0;
        int totalSuccessRule = 0;
        for (int x=0; x<mValidatorList.size(); x++){
            Validator validator = mValidatorList.get(x);
            totalRule += validator.mRuleList.size();
            for (int y=0; y<validator.mRuleList.size(); y++){
                AbstractValidatorRule rule = validator.mRuleList.get(y);
                if(validator.mView instanceof TextView) {
                    TextView mTextView = (TextView) validator.mView;
                    boolean isValid = rule.isValid(validator.mView, mTextView.getText().toString(), validator.packet);
                    if (!isValid) {
                        if(onValidationCallback.onFailed(validator.mView, rule, rule.getMessage()))break;
                    } else {
                        totalSuccessRule++;
                        if(onValidationCallback.onSuccess(validator.mView, rule))break;
                    }
                }
            }
        }
        onValidationCallback.onComplete(null, totalRule == totalSuccessRule);
        return totalRule == totalSuccessRule;
    }


    //Validator class
    public static class Validator{
        private Context mContext;
        private View mView;
        private Object packet;
        private List<AbstractValidatorRule> mRuleList;
        private OnValidationCallback onValidationCallback;
        private boolean alwaysShowErrorOnView = false;

        private void setup(Context context, View mView, Object packet){
            this.mContext = context.getApplicationContext();
            this.mView = mView;
            this.packet = packet;
            mRuleList = new ArrayList<>();
        }

        public AbstractValidatorRule getRule(int position){
            return mRuleList.get(position);
        }

        public int getRuleCount(){
            return mRuleList.size();
        }

        public void setPacket(Object packet) {
            this.packet = packet;
        }

        public Validator(Context context, View mView, Object packet){
            setup(context,mView,packet);
        }

        public Validator(Context context, int viewResourceId, Object packet){
            View editText = ((Activity)context).findViewById(viewResourceId);
            setup(context,editText,packet);
        }

        public Validator(Context context, View mView){
            setup(context,mView,null);
        }

        public Validator(Context context, int viewResourceId){
            View editText = ((Activity)context).findViewById(viewResourceId);
            setup(context,editText,null);
        }

        public Validator addValidatorRule(AbstractValidatorRule validatorRule){
            mRuleList.add(validatorRule);
            return this;
        }

        public Validator setOnValidationCallback(OnValidationCallback onValidationCallback, boolean alwaysShowErrorOnView) {
            this.onValidationCallback = onValidationCallback;
            this.alwaysShowErrorOnView = alwaysShowErrorOnView;
            return this;
        }

        public Validator setOnValidationCallback(OnValidationCallback onValidationCallback){
            setOnValidationCallback(onValidationCallback, false);
            return this;
        }

        public void validate(){
            int successCount = 0;
            for (int y=0; y<mRuleList.size(); y++){
                AbstractValidatorRule rule = mRuleList.get(y);
                if(mView instanceof TextView) {
                    TextView mTextView = (TextView) mView;
                    boolean isValid = false;
                    try {
                        isValid = rule.isValid(mView, mTextView.getText().toString(), packet);
                    } catch (ValidatorException e) {
                        e.printStackTrace();
                    }
                    if (!isValid) {
                        if(onValidationCallback != null){
                            if(alwaysShowErrorOnView){
                                mTextView.setError(rule.getMessage());
                            }
                            if(onValidationCallback.onFailed(mView, rule, rule.getMessage()))break;
                        } else {
                            mTextView.setError(rule.getMessage());
                        }
                    } else {
                        if(onValidationCallback != null){
                            successCount++;
                            if(alwaysShowErrorOnView){
                                mTextView.setError(null);
                            }
                            if(onValidationCallback.onSuccess(mView, rule))break;
                        } else {
                            mTextView.setError(null);
                        }
                    }
                }
            }
            if(onValidationCallback != null){
                onValidationCallback.onComplete(mView, successCount == mRuleList.size());
            }
        }
    }

    public interface OnValidationCallback{
        public boolean onSuccess(View view, AbstractValidatorRule validatorRule);
        public boolean onFailed(View view, AbstractValidatorRule validatorRule, String message);
        public void onComplete(View view, boolean allRuleValid);
    }

    public static abstract class AbstractValidatorRule{
        private Context mContext;
        private String mErrorMessage;
        public AbstractValidatorRule(Context context){
            this.mContext = context.getApplicationContext();
        }
        public AbstractValidatorRule(Context context, String errorMessage){
            this.mContext = context.getApplicationContext();
            this.mErrorMessage = errorMessage;
        }
        public abstract boolean isValid(View view, String value, Object packet) throws ValidatorException;


        public String getMessage(){
            return mErrorMessage;
        }

        public Context getContext() {
            return mContext;
        }

        public void setContext(Context mContext) {
            this.mContext = mContext;
        }

        public void setErrorMessage(String mErrorMessage) {
            this.mErrorMessage = mErrorMessage;
        }
    }



    //Validation Rules definition
    public static class NotEmptyValidatorRule extends AbstractValidatorRule{
        public NotEmptyValidatorRule(Context context) {
            super(context);
            setErrorMessage("Please fill this");
        }

        public NotEmptyValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            return !TextUtils.isEmpty(value);
        }
    }

    public static class EmailValidatorRule extends AbstractValidatorRule{
        private String mDomainName = "";
        public EmailValidatorRule(Context context) {
            super(context);
            setErrorMessage("Email is invalid");
        }

        public EmailValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if (!TextUtils.isEmpty(value)) {
                if (TextUtils.isEmpty(mDomainName)) {
                    Pattern pattern = Pattern.compile(".+@.+\\.[a-z]+");
                    Matcher matcher = pattern.matcher(value);
                    return matcher.matches();
                } else {
                    Pattern pattern = Pattern.compile(".+@" + mDomainName);
                    Matcher matcher = pattern.matcher(value);
                    return matcher.matches();
                }
            }else{
                return true;
            }
        }

        public void setDomainName(String domainName) {
            mDomainName = domainName;
        }
    }


    public static class RegExpValidatorRule extends AbstractValidatorRule{
        private Pattern mPattern;
        public RegExpValidatorRule(Context context) {
            super(context);
            setErrorMessage("Please insert with the valid format");
        }

        public RegExpValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if (mPattern != null) {
                return mPattern.matcher(value).matches();
            }
            throw new ValidatorException("You need to set Regexp Pattern first");
        }

        public void setPattern(String pattern) {
            mPattern = Pattern.compile(pattern);
        }

        public void setPattern(Pattern pattern) {
            mPattern = pattern;
        }
    }


    public static class AlphaNumericValidatorRule extends AbstractValidatorRule{
        public AlphaNumericValidatorRule(Context context) {
            super(context);
            setErrorMessage("Please fill with alpha numeric only");
        }

        public AlphaNumericValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if(TextUtils.isEmpty(value))return true;
            return TextUtils.isDigitsOnly(value);
        }
    }

    public static class NumericValidatorRule extends AbstractValidatorRule{
        public NumericValidatorRule(Context context) {
            super(context);
            setErrorMessage("Please fill with numeric only");
        }

        public NumericValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if(TextUtils.isEmpty(value))return true;
            if (value == null || value.length() == 0)
                return false;
            for (int i = 0; i < value.length(); i++) {
                if (!Character.isDigit(value.charAt(i)))
                    return false;
            }
            return true;
        }
    }

    public static class SameValueValidatorRule extends AbstractValidatorRule{
        private String comparedFieldName;
        public SameValueValidatorRule(Context context, String comparedFieldName) {
            super(context);
            this.comparedFieldName = comparedFieldName;
            setErrorMessage("The value is different with "+comparedFieldName);
        }

        public SameValueValidatorRule(Context context, String errorMessage, String comparedFieldName) {
            super(context, errorMessage);
            this.comparedFieldName = comparedFieldName;
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            View comparedView = (View)packet;
            String comparedValue = ((TextView)comparedView).getText().toString();
            return value.equals(comparedValue);
        }
    }

    public static class CountValidatorRule extends AbstractValidatorRule{
        private int count;
        public CountValidatorRule(Context context, int count) {
            super(context);
            setErrorMessage("Data must be "+count+(count>1?" digits":" digit"));
            this.count = count;
        }

        public CountValidatorRule(Context context, String errorMessage, int count) {
            super(context, errorMessage);
            this.count = count;
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            return value.length() == count;
        }
    }

    public static class PhoneValidatorRule extends AbstractValidatorRule{
        public PhoneValidatorRule(Context context) {
            super(context);
            setErrorMessage("Phone format is invalid");
        }

        public PhoneValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if(TextUtils.isEmpty(value))return true;
            return Patterns.PHONE.matcher(value).matches();
        }
    }

    public static class MustCheckedValidatorRule extends AbstractValidatorRule{
        public MustCheckedValidatorRule(Context context) {
            super(context);
            setErrorMessage("You must check this");
        }

        public MustCheckedValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if(view instanceof CheckBox){
                if(((CheckBox)view).isChecked()){
                    return true;
                }else{
                    //invalid state :
                    //1. remove error message after checked
                    ((CheckBox)view).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                        @Override
                        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                            if(isChecked){
                                buttonView.setError(null);
                                buttonView.setFocusableInTouchMode(false);
                            }
                        }
                    });
                    ((CheckBox)view).setFocusableInTouchMode(true);

                }
            }
            return false;
        }
    }


    public static class URLValidatorRule extends AbstractValidatorRule{
        public URLValidatorRule(Context context) {
            super(context);
            setErrorMessage("Url format is invalid");
        }

        public URLValidatorRule(Context context, String errorMessage) {
            super(context, errorMessage);
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            if(TextUtils.isEmpty(value))return true;
            return Patterns.WEB_URL.matcher(value).matches();
        }
    }

    public static class DateValidatorRule extends AbstractValidatorRule{
        private String dateFormat;
        private Locale locale;
        public DateValidatorRule(Context context, String dateFormat, Locale locale) {
            super(context);
            setErrorMessage("Date format is invalid");
            this.dateFormat = dateFormat;
            this.locale = locale;
        }

        public DateValidatorRule(Context context, String errorMessage, String dateFormat, Locale locale) {
            super(context, errorMessage);
            this.dateFormat = dateFormat;
            this.locale = locale;
        }

        @Override
        public boolean isValid(View view, String value, Object packet) throws ValidatorException {
            try {
                DateFormat df = new SimpleDateFormat(dateFormat, locale);
                df.setLenient(false);
                df.parse(value);
                return true;
            } catch (ParseException e) {
                return false;
            }
        }
    }



    public static class ValidatorException extends java.lang.Exception {
        public ValidatorException() {
            super();
        }

        public ValidatorException(String detailMessage, Throwable throwable) {
            super(detailMessage, throwable);
        }

        public ValidatorException(String detailMessage) {
            super(detailMessage);
        }

        public ValidatorException(Throwable throwable) {
            super(throwable);
        }
    }


    //helper method

    public Validator addNotEmptyValidatorForView(Context context, View view){
        Validator newValidator = new FormValidationUtils.Validator(context, view);
        newValidator.addValidatorRule(new NotEmptyValidatorRule(context));
        addValidator(newValidator);
        return newValidator;
    }

    public Validator addNotEmptyValidatorForView(Context context, View view, String errorMessage){
        Validator newValidator = new FormValidationUtils.Validator(context, view);
        newValidator.addValidatorRule(new NotEmptyValidatorRule(context, errorMessage));
        addValidator(newValidator);
        return newValidator;
    }

    public Validator addEmailValidatorForView(Context context, View view){
        Validator newValidator = new FormValidationUtils.Validator(context, view);
        newValidator.addValidatorRule(new EmailValidatorRule(context));
        addValidator(newValidator);
        return newValidator;
    }

    public Validator addEmailValidatorForView(Context context, View view, String errorMessage){
        Validator newValidator = new FormValidationUtils.Validator(context, view);
        newValidator.addValidatorRule(new EmailValidatorRule(context, errorMessage));
        addValidator(newValidator);
        return newValidator;
    }
}

File: zlcore/src/main/java/com/zaitunlabs/zlcore/utils/FormValidationUtils.kt
package com.zaitunlabs.zlcore.utils

import android.app.Activity
import android.content.Context
import android.text.TextUtils
import android.util.Patterns
import android.view.View
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.EditText
import android.widget.TextView

import com.zaitunlabs.zlcore.R

import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Locale
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * Created by ahsai on 11/21/2017.
 */

class FormValidationUtils(context: Context) {
    private val context: Context
    private val mValidatorList: MutableList<Validator>

    val validatorCount: Int
        get() = mValidatorList.size

    init {
        this.context = context.applicationContext
        mValidatorList = ArrayList()
    }

    @JvmOverloads
    fun addValidator(validator: Validator, validationType: Int = DefaultType) {
        mValidatorList.add(validator)
        if (validationType == DefaultType) {
        } else if (validationType == IdleType) {
            enableTypingValidation(validator)
        } else if (validationType == UnfocusType) {
            enableUnfocusValidation(validator)
        }
    }

    private fun enableUnfocusValidation(validator: Validator) {
        CommonUtils.performTaskWhenUnFocus(validator.mView) { validator.validate() }
    }

    private fun enableTypingValidation(validator: Validator) {
        if (validator.mView is EditText) {
            CommonUtils.performTaskWhenTypeIdle((validator.mView as EditText?)!!) { validator.validate() }
        }
    }

    fun getValidator(position: Int): Validator? {
        return if (position < mValidatorList.size) {
            mValidatorList[position]
        } else null
    }

    fun removeValidator(validator: Validator) {
        mValidatorList.remove(validator)
    }

    @Throws(FormValidationUtils.ValidatorException::class)
    fun validate(): Boolean {
        var isFirsErrorFocused = false
        for (x in mValidatorList.indices) {
            val validator = mValidatorList[x]
            for (y in validator.mRuleList!!.indices) {
                val rule = validator.mRuleList!![y]
                if (validator.mView is TextView) {
                    val mTextView = validator.mView as TextView?
                    val isValid = rule.isValid(validator.mView, mTextView!!.text.toString(), validator.packet)
                    if (!isValid) {
                        mTextView.error = rule.message
                        if (!isFirsErrorFocused) {
                            mTextView.requestFocus()
                            isFirsErrorFocused = true
                        }
                        break
                    } else {
                        mTextView.error = null
                    }
                }
            }
        }

        return !isFirsErrorFocused
    }

    @Throws(FormValidationUtils.ValidatorException::class)
    fun validate(onValidationCallback: OnValidationCallback): Boolean {
        var totalRule = 0
        var totalSuccessRule = 0
        for (x in mValidatorList.indices) {
            val validator = mValidatorList[x]
            totalRule += validator.mRuleList!!.size
            for (y in validator.mRuleList!!.indices) {
                val rule = validator.mRuleList!![y]
                if (validator.mView is TextView) {
                    val mTextView = validator.mView as TextView?
                    val isValid = rule.isValid(validator.mView, mTextView!!.text.toString(), validator.packet)
                    if (!isValid) {
                        if (onValidationCallback.onFailed(validator.mView, rule, rule.message)) break
                    } else {
                        totalSuccessRule++
                        if (onValidationCallback.onSuccess(validator.mView, rule)) break
                    }
                }
            }
        }
        onValidationCallback.onComplete(null, totalRule == totalSuccessRule)
        return totalRule == totalSuccessRule
    }


    //Validator class
    class Validator {
        private var mContext: Context? = null
        private var mView: View? = null
        private var packet: Any? = null
        private var mRuleList: MutableList<AbstractValidatorRule>? = null
        private var onValidationCallback: OnValidationCallback? = null
        private var alwaysShowErrorOnView = false

        val ruleCount: Int
            get() = mRuleList!!.size

        private fun setup(context: Context, mView: View, packet: Any?) {
            this.mContext = context.applicationContext
            this.mView = mView
            this.packet = packet
            mRuleList = ArrayList()
        }

        fun getRule(position: Int): AbstractValidatorRule {
            return mRuleList!![position]
        }

        fun setPacket(packet: Any) {
            this.packet = packet
        }

        constructor(context: Context, mView: View, packet: Any) {
            setup(context, mView, packet)
        }

        constructor(context: Context, viewResourceId: Int, packet: Any) {
            val editText = (context as Activity).findViewById<View>(viewResourceId)
            setup(context, editText, packet)
        }

        constructor(context: Context, mView: View) {
            setup(context, mView, null)
        }

        constructor(context: Context, viewResourceId: Int) {
            val editText = (context as Activity).findViewById<View>(viewResourceId)
            setup(context, editText, null)
        }

        fun addValidatorRule(validatorRule: AbstractValidatorRule): Validator {
            mRuleList!!.add(validatorRule)
            return this
        }

        fun setOnValidationCallback(onValidationCallback: OnValidationCallback, alwaysShowErrorOnView: Boolean): Validator {
            this.onValidationCallback = onValidationCallback
            this.alwaysShowErrorOnView = alwaysShowErrorOnView
            return this
        }

        fun setOnValidationCallback(onValidationCallback: OnValidationCallback): Validator {
            setOnValidationCallback(onValidationCallback, false)
            return this
        }

        fun validate() {
            var successCount = 0
            for (y in mRuleList!!.indices) {
                val rule = mRuleList!![y]
                if (mView is TextView) {
                    val mTextView = mView as TextView?
                    var isValid = false
                    try {
                        isValid = rule.isValid(mView, mTextView!!.text.toString(), packet)
                    } catch (e: ValidatorException) {
                        e.printStackTrace()
                    }

                    if (!isValid) {
                        if (onValidationCallback != null) {
                            if (alwaysShowErrorOnView) {
                                mTextView!!.error = rule.message
                            }
                            if (onValidationCallback!!.onFailed(mView, rule, rule.message)) break
                        } else {
                            mTextView!!.error = rule.message
                        }
                    } else {
                        if (onValidationCallback != null) {
                            successCount++
                            if (alwaysShowErrorOnView) {
                                mTextView!!.error = null
                            }
                            if (onValidationCallback!!.onSuccess(mView, rule)) break
                        } else {
                            mTextView!!.error = null
                        }
                    }
                }
            }
            if (onValidationCallback != null) {
                onValidationCallback!!.onComplete(mView, successCount == mRuleList!!.size)
            }
        }
    }

    interface OnValidationCallback {
        fun onSuccess(view: View?, validatorRule: AbstractValidatorRule): Boolean
        fun onFailed(view: View?, validatorRule: AbstractValidatorRule, message: String?): Boolean
        fun onComplete(view: View?, allRuleValid: Boolean)
    }

    abstract class AbstractValidatorRule {
        var context: Context? = null
        var message: String? = null
            private set

        constructor(context: Context) {
            this.context = context.applicationContext
        }

        constructor(context: Context, errorMessage: String) {
            this.context = context.applicationContext
            this.message = errorMessage
        }

        @Throws(FormValidationUtils.ValidatorException::class)
        abstract fun isValid(view: View?, value: String, packet: Any?): Boolean

        fun setErrorMessage(mErrorMessage: String) {
            this.message = mErrorMessage
        }
    }


    //Validation Rules definition
    class NotEmptyValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_please_fill_this))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            return !TextUtils.isEmpty(value)
        }
    }

    class EmailValidatorRule : AbstractValidatorRule {
        private var mDomainName = ""

        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_email_is_invalid))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            if (!TextUtils.isEmpty(value)) {
                if (TextUtils.isEmpty(mDomainName)) {
                    val pattern = Pattern.compile(".+@.+\\.[a-z]+")
                    val matcher = pattern.matcher(value)
                    return matcher.matches()
                } else {
                    val pattern = Pattern.compile(".+@$mDomainName")
                    val matcher = pattern.matcher(value)
                    return matcher.matches()
                }
            } else {
                return true
            }
        }

        fun setDomainName(domainName: String) {
            mDomainName = domainName
        }
    }


    class RegExpValidatorRule : AbstractValidatorRule {
        private var mPattern: Pattern? = null

        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_please_insert_with_valid_format))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            if (mPattern != null) {
                return mPattern!!.matcher(value).matches()
            }
            throw ValidatorException(view!!.context.getString(R.string.zlcore_form_validations_utils_need_set_regex_pattern))
        }

        fun setPattern(pattern: String) {
            mPattern = Pattern.compile(pattern)
        }

        fun setPattern(pattern: Pattern) {
            mPattern = pattern
        }
    }


    class AlphaNumericValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_please_fill_alpha_numeric_only))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            return if (TextUtils.isEmpty(value)) true else TextUtils.isDigitsOnly(value)
        }
    }

    class NumericValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_please_fill_numeric_only))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String?, packet: Any?): Boolean {
            if (TextUtils.isEmpty(value)) return true
            if (value == null || value.length == 0)
                return false
            for (i in 0 until value.length) {
                if (!Character.isDigit(value[i]))
                    return false
            }
            return true
        }
    }

    class SameValueValidatorRule : AbstractValidatorRule {
        private var comparedFieldName: String? = null

        constructor(context: Context, comparedFieldName: String) : super(context) {
            this.comparedFieldName = comparedFieldName
            setErrorMessage(String.format(context.getString(R.string.zlcore_form_validations_utils_value_is_different), comparedFieldName))
        }

        constructor(context: Context, errorMessage: String, comparedFieldName: String) : super(context, errorMessage) {
            this.comparedFieldName = comparedFieldName
        }

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            val comparedView = packet as View?
            val comparedValue = (comparedView as TextView).text.toString()
            return value == comparedValue
        }
    }

    class CountValidatorRule : AbstractValidatorRule {
        private var count: Int = 0

        constructor(context: Context, count: Int) : super(context) {
            setErrorMessage(String.format(context.getString(R.string.zlcore_form_validations_utils_data_must_be_count_digits), count, if (count > 1) "digits" else "digit"))
            this.count = count
        }

        constructor(context: Context, errorMessage: String, count: Int) : super(context, errorMessage) {
            this.count = count
        }

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            return value.length == count
        }
    }

    class PhoneValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_phone_format_is_invalid))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            return if (TextUtils.isEmpty(value)) true else Patterns.PHONE.matcher(value).matches()
        }
    }

    class MustCheckedValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_you_must_check_this))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            if (view is CheckBox) {
                if (view.isChecked) {
                    return true
                } else {
                    //invalid state :
                    //1. remove error message after checked
                    view.setOnCheckedChangeListener { buttonView, isChecked ->
                        if (isChecked) {
                            buttonView.error = null
                            buttonView.isFocusableInTouchMode = false
                        }
                    }
                    view.isFocusableInTouchMode = true

                }
            }
            return false
        }
    }


    class URLValidatorRule : AbstractValidatorRule {
        constructor(context: Context) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_url_format_is_invalid))
        }

        constructor(context: Context, errorMessage: String) : super(context, errorMessage) {}

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            return if (TextUtils.isEmpty(value)) true else Patterns.WEB_URL.matcher(value).matches()
        }
    }

    class DateValidatorRule : AbstractValidatorRule {
        private var dateFormat: String? = null
        private var locale: Locale? = null

        constructor(context: Context, dateFormat: String, locale: Locale) : super(context) {
            setErrorMessage(context.getString(R.string.zlcore_form_validations_utils_date_format_is_invalid))
            this.dateFormat = dateFormat
            this.locale = locale
        }

        constructor(context: Context, errorMessage: String, dateFormat: String, locale: Locale) : super(context, errorMessage) {
            this.dateFormat = dateFormat
            this.locale = locale
        }

        @Throws(FormValidationUtils.ValidatorException::class)
        override fun isValid(view: View?, value: String, packet: Any?): Boolean {
            try {
                val df = SimpleDateFormat(dateFormat, locale)
                df.isLenient = false
                df.parse(value)
                return true
            } catch (e: ParseException) {
                return false
            }

        }
    }


    class ValidatorException : java.lang.Exception {
        constructor() : super() {}

        constructor(detailMessage: String, throwable: Throwable) : super(detailMessage, throwable) {}

        constructor(detailMessage: String) : super(detailMessage) {}

        constructor(throwable: Throwable) : super(throwable) {}
    }


    //helper method

    fun addNotEmptyValidatorForView(context: Context, view: View): Validator {
        val newValidator = FormValidationUtils.Validator(context, view)
        newValidator.addValidatorRule(NotEmptyValidatorRule(context))
        addValidator(newValidator)
        return newValidator
    }

    fun addNotEmptyValidatorForView(context: Context, view: View, errorMessage: String): Validator {
        val newValidator = FormValidationUtils.Validator(context, view)
        newValidator.addValidatorRule(NotEmptyValidatorRule(context, errorMessage))
        addValidator(newValidator)
        return newValidator
    }

    fun addEmailValidatorForView(context: Context, view: View): Validator {
        val newValidator = FormValidationUtils.Validator(context, view)
        newValidator.addValidatorRule(EmailValidatorRule(context))
        addValidator(newValidator)
        return newValidator
    }

    fun addEmailValidatorForView(context: Context, view: View, errorMessage: String): Validator {
        val newValidator = FormValidationUtils.Validator(context, view)
        newValidator.addValidatorRule(EmailValidatorRule(context, errorMessage))
        addValidator(newValidator)
        return newValidator
    }

    companion object {
        val DefaultType = 1
        val IdleType = 2
        val UnfocusType = 3
    }
}
